CRM/Mailing/BAO/Mailing.php

```diff
@@ -109,46 +109,227 @@ public static function getRecipientsCount($job_id, $mailing_id = NULL) {
+      return;
+    }
+
+    list($location_filter, $order_by) = self::getLocationFilterAndOrderBy($mailingObj->email_selection_method, $mailingObj->location_type_id);
+
+    // get all the saved searches AND hierarchical groups
+    // and load them in the cache
+    foreach ($recipientsGroup as $groupType => $groupIDs) {
+      $sql = sprintf("
+       SELECT *
+       FROM  civicrm_group g
+       WHERE id IN (%s) AND ( saved_search_id != 0 OR saved_search_id IS NOT NULL OR children IS NOT NULL )
+      ", implode(',', $groupIDs));
+
+      $groupDAO = CRM_Core_DAO::executeQuery($sql);
+      while ($groupDAO->fetch()) {
+        if ($groupDAO->cache_date == NULL) {
+          CRM_Contact_BAO_GroupContactCache::load($groupDAO);
+        }
+        if ($groupType == 'Include') {
+          $includeSmartGroupIDs[] = $groupDAO->id;
+        }
+        else {
+          $excludeSmartGroupIDs[] = $groupDAO->id;
+        }
+      }
+    }
+
+    // Create a temp table for contact exclusion.
+    $excludeTempTablename = "excluded_recipients_mailing_temp";
+    $includedTempTablename = "included_recipients_mailing_temp";
+    $mailingGroup->query(
+      "CREATE TEMPORARY TABLE $excludeTempTablename
+            (contact_id int primary key)
+            ENGINE=HEAP"
+    );
+    if (!empty($recipientsGroup['Exclude'])) {
+      $sql = sprintf("
+        INSERT INTO %s (contact_id)
+          SELECT DISTINCT gc.contact_id
+          FROM  civicrm_group_contact gc
+          WHERE gc.status = 'Added' AND gc.id IN (%s)", $excludeTempTablename, implode(',', $recipientsGroup['Exclude']));
+      $mailingGroup->query($sql);
+
+      if (count($excludeSmartGroupIDs)) {
+        $sql = sprintf("
+          INSERT IGNORE INTO %s (contact_id)
+            SELECT c.contact_id
+            FROM   civicrm_group_contact_cache c
+            WHERE  c.group_id IN (%s)
+        ", $excludeTempTablename, implode(',', $excludeSmartGroupIDs));
+        $mailingGroup->query($sql);
+      }
+    }
+
+    if (!empty($recipientsGroup['Base'])) {
+      $sql = sprintf("
+        INSERT IGNORE INTO %s (contact_id)
+          SELECT DISTINCT contact_id
+          FROM  civicrm_group_contact gc
+          WHERE status = 'Removed' AND group_id IN (%s)", $excludeTempTablename, implode(',', $recipientsGroup['Base']));
+      $mailingGroup->query($sql);
+    }
+
+    // Add all the (intended) recipients of an excluded prior mailing to
+    // the temp table.
+    $excludeSubMailing = sprintf("
+      INSERT IGNORE INTO %s (contact_id)
+        SELECT DISTINCT meq.contact_id
+        FROM  civicrm_mailing_event_queue meq
+         INNER JOIN  civicrm_mailing_job mj ON meq.job_id = mj.id
+         INNER JOIN  civicrm_mailing_group mg ON mj.mailing_id = mg.entity_id AND mg.entity_table = 'civicrm_mailing'
+        WHERE  mg.mailing_id = %d AND mg.group_type = 'Exclude'", $excludeTempTablename, $mailingID);
```

Comment:
I'm probably confused, but isn't $mailingID the ID for the current mailing, so the comment on L211 doesn't make sense to me. If $mailingID is the current mailing, then the only reason this works is that the full Mailing Event Queue has not yet been built.